ENUM metadata support, treating them as scalars rather than using a container type

azrogers opened the pull request:
This is an alternative version of #1085, implementing enums by leveraging our existing scalar metadata support rather than implementing them as an entirely new container type. This requires some changes to the assumptions we're currently making in the metadata system, chief among them the breaking of the connection between C++ type and `PropertyType`. However, it does simplify the implementation and reduce the amount of copy-and-paste necessary to support enums. Please review and discuss which approach you prefer!

kring commented on CesiumGltf/include/CesiumGltf/PropertyTableView.h:
```diff
@@ -1095,19 +1127,43 @@ class PropertyTableView {
+        classProperty.componentType.value_or(""));
+    const CesiumGltf::Enum* pEnumDefinition = nullptr;
+    if (type == PropertyType::Enum && classProperty.enumType &&
+        !IsMetadataBoolean<T>::value) {
```
Why is this check for boolean needed here?

azrogers replied:
Because we can't have a boolean enum. So if the type is `bool` and the property type is `Enum`, we will go to the else if branch where we return an `ErrorInvalidEnum` status.

j9liu replied:
I'm also a bit confused, would that not get caught by `!canRepresentPropertyType<T>(type)`? (And if not, maybe the function should be changed to reflect this?)

kring commented on CesiumGltf/test/TestPropertyAttributePropertyView.cpp:
```diff
@@ -1,3 +1,5 @@
+#include "CesiumNativeTests/TypeToPropertyTypeString.h"
```
Angle brackets here and in the other files.

kring reviewed (commented):
I have a few comments below, but overall this looks reasonable to me. I'm going to look at the Unreal implementation next, though, so that could change my opinion!

azrogers commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -153,25 +161,91 @@ validatePropertyType(const ClassProperty& classProperty) {
-      convertStringToPropertyType(classProperty.type)) {
+PropertyViewStatusType validatePropertyType(
+    const ClassProperty& classProperty,
+    const CesiumGltf::Enum* pEnumDefinition) {
```
Went with adding a null check instead, since a nullptr enum definition should be an invalid enum status.

azrogers commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -153,25 +161,91 @@ validatePropertyType(const ClassProperty& classProperty) {
+validateArrayPropertyType(const ClassProperty& classProperty) {
+  using ElementType = typename MetadataArrayType<T>::type;
+
+  if (convertStringToPropertyType(classProperty.type) != PropertyType::Enum) {
```
Originally I was just skipping the parts of validation that wouldn't work for enums, then I changed my mind and made a separate validation function for enums instead. Just forgot to remove this part.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -1435,6 +1670,66 @@ class PropertyView<PropertyArrayView<ElementType>, false> {
   }
 
+  /**
+   * @brief Constructs a property instance from a class definition only.
```
```suggestion
   * @brief Constructs a property instance from a class and enum definition.
```

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTexturePropertyView.h:
```diff
@@ -386,6 +385,85 @@ class PropertyTexturePropertyView<ElementType, false>
+   * @param image The {@link ImageAsset} used by the property.
+   * @param options The options for constructing the view.
+   */
+  PropertyTexturePropertyView(
```
If this is largely copy and paste, can the other constructor reference this one with a `nullptr` pEnumDefinition?

azrogers replied:
Unfortunately, no - this constructor calls out to the `PropertyView` constructor which calls out to `validatePropertyType`. `validatePropertyType` with the enum overload will return `ErrorInvalidEnum` with a nullptr enum definition (as it should).

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTablePropertyView.h:
```diff
@@ -258,6 +259,34 @@ class PropertyTablePropertyView<ElementType, false>
+      const ClassProperty& classProperty,
+      int64_t size,
+      std::span<const std::byte> values,
+      const CesiumGltf::Enum* pEnumDefinition) noexcept
```
Sorry, this is my OCD talking -- it feels inconsistent that the `pEnumDefinition` is at the bottom for `PropertyTablePropertyView` , but comes right after `ClassProperty` in `PropertyTexturePropertyView` . Can we pick one? :')

j9liu reviewed (changes requested):
@azrogers I haven't done a full review, but I at least wanted to check the areas that Kevin commented on. I'll do another pass early next week

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTypeTraits.h:
```diff
@@ -201,9 +201,12 @@ struct MetadataArrayType<CesiumGltf::PropertyArrayCopy<T>> {
-/**
- * @brief Convert a C++ type to PropertyType and PropertyComponentType
- */
+/** * @brief Infer the best-fitting PropertyType and PropertyComponentType for a
```
Something went wrong with the asterisks in this comment...

j9liu commented on CesiumGltf/test/TestPropertyTypeTraits.cpp:
```diff
@@ -193,365 +193,65 @@ TEST_CASE("Test MetadataArrayType") {
   REQUIRE(std::is_same_v<type, uint32_t>);
 }
 
-TEST_CASE("TypeToPropertyType") {
```
Now that we're preserving `TypeToPropertyType`, we should preserve the unit tests for it too.

j9liu commented on CesiumGltf/test/TestPropertyView.cpp:
```diff
@@ -3058,3 +3187,215 @@ TEST_CASE("String Array PropertyView") {
   }
 }
+
+TEST_CASE("Enum PropertyView") {
```
Sorry for the nitpick, but can this be put under the `String PropertyView` tests? Just so the non-array tests are grouped together.

j9liu commented on CesiumGltf/test/TestPropertyView.cpp:
```diff
@@ -3058,3 +3187,215 @@ TEST_CASE("String Array PropertyView") {
+  SUBCASE("Correct property type") {
+    ClassProperty classProperty;
+    classProperty.type = ClassProperty::Type::ENUM;
+    classProperty.array = true;
+    PropertyView<PropertyArrayView<uint8_t>> view(classProperty);
+    REQUIRE(view.propertyType() == PropertyType::Enum);
```
```suggestion
    PropertyView<uint8_t> view(classProperty);
    REQUIRE(view.propertyType() == PropertyType::Enum);
```

j9liu commented on CesiumGltf/test/TestPropertyView.cpp:
```diff
@@ -3058,3 +3187,215 @@ TEST_CASE("String Array PropertyView") {
+    PropertyView<uint8_t> view(classProperty, &enumDef);
+    REQUIRE(view.status() == PropertyViewStatus::ErrorInvalidDefaultValue);
+
+    classProperty.noData = "null";
+    view = PropertyView<uint8_t>(classProperty, &enumDef);
+    REQUIRE(view.status() == PropertyViewStatus::ErrorInvalidNoDataValue);
+
+    classProperty.noData = "Qux";
+    view = PropertyView<uint8_t>(classProperty, &enumDef);
+    REQUIRE(view.status() == PropertyViewStatus::ErrorInvalidNoDataValue);
+  }
```
Pick one? These seem like they're testing the same thing, just with different random strings

azrogers replied:
They're not! The "null" string is interpreted as a null JSON value, which tests that the `noData` handling correctly ignores non-string values. The "Qux" string, on the other hand, is a JSON string that isn't in the enum definition, so it's testing that those types of invalid `noData` values will also be handled correctly.

j9liu replied:
I don't think that's right, because we have this test for string property views:

```
    ClassProperty classProperty;
    classProperty.type = ClassProperty::Type::STRING;
    classProperty.required = false;
    classProperty.noData = "null";
    classProperty.defaultProperty = "default";

    PropertyView<std::string_view> view(classProperty);
    REQUIRE(view.status() == PropertyViewStatus::Valid);
    REQUIRE(!view.required());

    REQUIRE(view.noData() == "null");
    REQUIRE(view.defaultValue() == "default");
```

If `noData` didn't actually exist -- i.e., was a null JSON value--it wouldn't pass that `REQUIRE(view.noData() == "null");` check.

The reason that this test is failing is actually because `required = true` in the class property. Required properties are not allowed to have a `noData` value. Hence why the unit tests end up being duplicate.

j9liu replied:
It sounds like we need a separate test to check whether the input `noData` / `default` values actually exist in the enum -- and yes, we should test that.

j9liu replied:
@azrogers I think you may have missed this comment

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -182,6 +225,39 @@ validateArrayPropertyType(const ClassProperty& classProperty) {
+ * validating, or \ref PropertyViewStatus::Valid if no errors were found.
+ */
+template <typename T>
+PropertyViewStatusType validateArrayPropertyType(
```
See comment on `PropertyTextureView`. I think the same consolidation could be applied here, too.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -381,6 +459,68 @@ template <typename ElementType> class PropertyView<ElementType, false> {
+   * @brief Constructs a property instance from a class definition and enum
+   * definition.
+   */
+  PropertyView(
```
I kinda feel the same way about the duplication here. I think allowing `pEnumDefinition = nullptr` and doing an extra branch for enums would reduce duplication. Assuming everything else can maintain correctness.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -145,6 +153,40 @@ validatePropertyType(const ClassProperty& classProperty) {
 }
 
+/**
+ * @brief Validates a \ref ClassProperty representing an array, checking for any
```
Think the doc is wrong here, this isn't an array type

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -417,6 +558,22 @@ template <typename ElementType> class PropertyView<ElementType, false> {
+   * @brief Constructs a property instance from a property table property, class
+   * definition, and an associated enum definition.
+   */
+  PropertyView(
```
Same with the rest of the duplicate constructors in this file. It would be great to consolidate as much as possible.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -565,6 +755,12 @@ template <typename ElementType> class PropertyView<ElementType, false> {
   }
 
+  /**
+   * @brief Returns the \ref PropertyType of the property this view is
```
```suggestion
   * @brief Returns the \ref PropertyType of the property that this view is
```

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTypeTraits.h:
```diff
@@ -392,6 +395,48 @@ template <> struct TypeToPropertyType<std::string_view> {
+/**
+ * @brief The number of dimensions that this type contains.
+ */
+template <typename T> struct TypeToDimensions;
```
Just wanted to say I like this change 😄 thanks for making it!

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -1435,6 +1678,67 @@ class PropertyView<PropertyArrayView<ElementType>, false> {
+      return;
+    }
+
+    getNumericPropertyValues(classProperty);
```
`getNumericPropertyValues` handles `offset`, `scale`, `max`, `min`, which don't apply to enum properties. When you consolidate the constructors, just check the property type here.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTablePropertyView.h:
```diff
@@ -258,6 +259,34 @@ class PropertyTablePropertyView<ElementType, false>
+   * @param classProperty The {@link ClassProperty} this property conforms to.
+   * @param size The number of elements in the property table specified by {@link PropertyTable::count}
+   * @param values The raw buffer specified by {@link PropertyTableProperty::values}
+   * @param pEnumDefinition A pointer to the enum definition used for this
```
Can `@param pEnumDefinition` be moved to reflect the new parameter order?

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTablePropertyView.h:
```diff
@@ -258,6 +259,34 @@ class PropertyTablePropertyView<ElementType, false>
+      const CesiumGltf::Enum* pEnumDefinition,
+      int64_t size,
+      std::span<const std::byte> values) noexcept
+      : PropertyView<ElementType>(classProperty, property, pEnumDefinition),
```
Could you consolidate this with the other constructor?

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTablePropertyView.h:
```diff
@@ -290,6 +319,41 @@ class PropertyTablePropertyView<ElementType, false>
+   * @param pEnumDefinition A pointer to the enum definition used for this
+   * value.
+   */
+  PropertyTablePropertyView(
```
Same here.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTableView.h:
```diff
@@ -194,13 +195,44 @@ class PropertyTableView {
+    if (type == PropertyType::Enum && pClassProperty->enumType) {
+      const auto& enumDefinitionIt =
+          this->_pEnumDefinitions->find(*pClassProperty->enumType);
+      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
+        callback(
+            propertyId,
+            PropertyTablePropertyView<uint8_t>(
+                PropertyTablePropertyViewStatus::ErrorInvalidEnum));
+        return;
+      } else {
+        componentType = convertStringToPropertyComponentType(
+            enumDefinitionIt->second.valueType);
+
+        if (componentType == PropertyComponentType::Float32 ||
+            componentType == PropertyComponentType::Float64) {
+          callback(
+              propertyId,
+              PropertyTablePropertyView<uint8_t>(
+                  PropertyTablePropertyViewStatus::ErrorInvalidEnum));
+          return;
+        }
+      }
```
Might be nitpicky, but can be flatten the `if else` here so there's less indentation? It helps with readability at the end of the day

azrogers replied:
I don't see how it can be flattened?

j9liu replied:
Sorry, I just meant the outer if/else statement. Like this:

```suggestion
      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
        callback(
            propertyId,
            PropertyTablePropertyView<uint8_t>(
                PropertyTablePropertyViewStatus::ErrorInvalidEnum));
        return;
      }
        componentType = convertStringToPropertyComponentType(
            enumDefinitionIt->second.valueType);

        if (componentType == PropertyComponentType::Float32 ||
            componentType == PropertyComponentType::Float64) {
          callback(
              propertyId,
              PropertyTablePropertyView<uint8_t>(
                  PropertyTablePropertyViewStatus::ErrorInvalidEnum));
          return;
        }
      
```

j9liu replied:
Same with all the others.

j9liu replied:
@azrogers and this one.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTableView.h:
```diff
@@ -1138,11 +1193,28 @@ class PropertyTableView {
-        classProperty,
-        _pPropertyTable->count,
-        values);
+    if constexpr (!IsMetadataBoolean<T>::value && !Normalized) {
+      if (type == PropertyType::Enum) {
+        return PropertyTablePropertyView<T, Normalized>(
+            propertyTableProperty,
+            classProperty,
+            pEnumDefinition,
+            _pPropertyTable->count,
+            values);
+      } else {
+        return PropertyTablePropertyView<T, Normalized>(
+            propertyTableProperty,
+            classProperty,
+            _pPropertyTable->count,
+            values);
+      }
+    } else {
+      return PropertyTablePropertyView<T, Normalized>(
+          propertyTableProperty,
+          classProperty,
+          _pPropertyTable->count,
+          values);
+    }
```
Will these different branches be needed after all the constructor changes?

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTextureView.h:
```diff
@@ -595,14 +596,37 @@ class PropertyTextureView {
+    if (type == PropertyType::Enum && classProperty.enumType) {
+      const auto& enumDefinitionIt =
+          this->_pEnumDefinitions->find(*classProperty.enumType);
+      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
+        return PropertyTexturePropertyView<T, Normalized>(
+            PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
+      } else {
+        componentType = convertStringToPropertyComponentType(
+            enumDefinitionIt->second.valueType);
+        pEnumDefinition = &enumDefinitionIt->second;
+
+        if (componentType == PropertyComponentType::Float32 ||
+            componentType == PropertyComponentType::Float64) {
+          return PropertyTexturePropertyView<T, Normalized>(
+              PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
+        }
+      }
```
Could we flatten this `if`/`else`?

azrogers replied:
Similarly, I don't see how this can be flattened. We can't do the lookup unless it's an enum value.

j9liu replied:
```suggestion
      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
        return PropertyTexturePropertyView<T, Normalized>(
            PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
      }
        componentType = convertStringToPropertyComponentType(
            enumDefinitionIt->second.valueType);
        pEnumDefinition = &enumDefinitionIt->second;

        if (componentType == PropertyComponentType::Float32 ||
            componentType == PropertyComponentType::Float64) {
          return PropertyTexturePropertyView<T, Normalized>(
              PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
        }
      
```

j9liu reviewed (changes requested):
@azrogers Thank you for responding quickly to feedback and putting effort into these changes!

My one remaining reservation is about the way that we handle the constructors with `pEnumDefinition`. There is a silent assumption that if we call the constructor that includes the enum definition parameter -- even if `nullptr` -- that the property must have been an enum. I feel like this isn't right; we could be handling the inputs more generally, starting with `PropertyView` then working its way up into `PropertyTablePropertyView`, etc.

Let me know what you think -- overall I think it's going to reduce duplication.

azrogers commented on CesiumGltf/include/CesiumGltf/PropertyTextureView.h:
```diff
@@ -680,14 +706,37 @@ class PropertyTextureView {
+    if (type == PropertyType::Enum && classProperty.enumType) {
+      const auto& enumDefinitionIt =
+          this->_pEnumDefinitions->find(*classProperty.enumType);
+      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
+        return PropertyTexturePropertyView<PropertyArrayView<T>, Normalized>(
+            PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
+      } else {
+        componentType = convertStringToPropertyComponentType(
+            enumDefinitionIt->second.valueType);
+        pEnumDefinition = &enumDefinitionIt->second;
+
+        if (componentType == PropertyComponentType::Float32 ||
+            componentType == PropertyComponentType::Float64) {
+          return PropertyTexturePropertyView<PropertyArrayView<T>, Normalized>(
+              PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
+        }
+      }
```
Same here, how would you flatten it?

j9liu replied:
```suggestion
      if (enumDefinitionIt == this->_pEnumDefinitions->end()) {
        return PropertyTexturePropertyView<PropertyArrayView<T>, Normalized>(
            PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
      } 
        componentType = convertStringToPropertyComponentType(
            enumDefinitionIt->second.valueType);
        pEnumDefinition = &enumDefinitionIt->second;

        if (componentType == PropertyComponentType::Float32 ||
            componentType == PropertyComponentType::Float64) {
          return PropertyTexturePropertyView<PropertyArrayView<T>, Normalized>(
              PropertyTexturePropertyViewStatus::ErrorInvalidEnum);
        }
      
```

azrogers commented:
@j9liu Addressed your review comments. Was able to consolidate most of the constructors - I did have to add a `pEnumDefinition` parameter to some of the protected constructors that would never be enums, just so the templating worked out, but since the parameter is ignored I don't think it's a big deal.

j9liu commented:
Thanks so much @azrogers! I'll take another look soon.

> Was able to consolidate most of the constructors - I did have to add a pEnumDefinition parameter to some of the protected constructors that would never be enums, just so the templating worked out, but since the parameter is ignored I don't think it's a big deal.

Yeah, this would usually irk me on a conceptual level, but we have a pattern of ignoring parameters anyway (like how `PropertyTableProperty` is ignored for strings and booleans). And I think the reduced duplication is worth it.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTableView.h:
```diff
@@ -1220,11 +1306,28 @@ class PropertyTableView {
-          classProperty,
-          _pPropertyTable->count,
-          values);
+      if constexpr (!IsMetadataBoolean<T>::value && !Normalized) {
```
We shouldn't be calling `getNumericArrayPropertyValues` for `T = bool` so I don't think this check is necessary

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyTableView.h:
```diff
@@ -1259,6 +1362,20 @@ class PropertyTableView {
           arrayOffsets,
           arrayOffsetType);
     } else {
+      if constexpr (!IsMetadataBoolean<T>::value) {
```
Same here.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -108,34 +111,59 @@ class PropertyViewStatus {
+ * @brief Validates a \ref ClassProperty representing a property, checking for
+ * any type mismatches.
+ *
+ * @tparam T The value type of the PropertyView we are validating.
```
I tried not to nitpick this since I doubt this function will be used by anyone else, but... I think it's better to use a neutral voice in the `/** */` documentation

```suggestion
 * @tparam T The value type of the PropertyView that was constructed for this ClassProperty.
```

azrogers replied:
What do you mean by "neutral voice?" If you mean the phrasing using "we," we're using that phrasing, by my count, at over 500 places in our documentation currently.

j9liu replied:
Yes I'm referring to the phrasing using "we", and personally I'd prefer avoiding it. I'm surprised by how much we are apparently using it in our in-code documentation.

It's different if we're talking through logic in a code body or an algorithm, but these are top-level comments 😕

azrogers replied:
I guess I just don't see the issue with using "we"? Not only do we use it in top-level comments in Cesium Native and Cesium for Unreal, but in CesiumJS as well. And my accidental ctrl-shift-f through the whole Cesium for Unreal VS solution says that "we" is used throughout the Unreal source code and even in the MSVC standard library. And I think it makes sense - if you were explaining an algorithm to someone, you say "First we do this step, then we do this second step, and then we take the result and do this to it."

azrogers replied:
I've changed the phrasing for the sake of getting the review done with, but I think if we want to decide on particular dos and don'ts for doc comments, that's something that should be in the style guide.

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -146,33 +174,52 @@ validatePropertyType(const ClassProperty& classProperty) {
+ * @brief Validates a \ref ClassProperty representing an array of values,
+ * checking for any type mismatches.
+ *
+ * @tparam T The array type of the PropertyView we are validating.
```
```suggestion
 * @tparam T The array element type of the PropertyView that was constructed for this ClassProperty.
```

azrogers replied:
`T` is the array type, not the element type.

j9liu replied:
Sorry I read too fast. Could you keep the rest of the wording without `element` added?

j9liu commented on CesiumGltf/include/CesiumGltf/PropertyView.h:
```diff
@@ -349,14 +408,20 @@ template <typename ElementType> class PropertyView<ElementType, false> {
 
     if (classProperty.noData) {
-      if (!_required) {
+      if (!_required && _propertyType == PropertyType::Enum &&
```
Could we rationalize that if we get to this point and `propertyType == Enum` that `pEnumDefinition` is also non-null, so we no longer have to do a `nullptr` check? (Maybe an `assert` would help, or prove me wrong? 😝 )

j9liu commented on CesiumGltf/test/TestPropertyTypeTraits.cpp:
```diff
@@ -193,275 +193,643 @@ TEST_CASE("Test MetadataArrayType") {
-    REQUIRE(TypeToPropertyType<uint8_t>::value == PropertyType::Scalar);
-    REQUIRE(
-        TypeToPropertyType<uint8_t>::component == PropertyComponentType::Uint8);
+TEST_CASE("Test canRepresentPropertyType") {
```
Sorry to trouble but do you mind ordering the tests starting from `TypeToPropertyType` -> `CanBeNormalized` -> `canRepresentPropertyType`? I'm sure there's nothing wrong with the content, but the visual diff is large on Github so it's a bit hard to compare

j9liu reviewed (commented):
Looking much better @azrogers ! :smile: I have a few remaining points but nothing major about the design. The constructors make everything much cleaner :smile:

Could you also make a note of this in CHANGES.md when you get the chance?

azrogers commented:
Sorry I missed those comments @j9liu, I believe I've addressed them all now.

azrogers commented:
@j9liu Ah, there it is, it was hidden under a "show more" that I hadn't clicked. I've removed the redundant part of the test and made a separate test for correct enum `noData` and `defaultProperty` values.

j9liu commented:
All good @azrogers , I know there's a lot 😁

Looks great now! Thanks for all the hard work on this PR.
